### PR TITLE
OAK-9325 Tool to compare index definitions

### PR DESCRIPTION
**Pull request by thomasmueller**

### CODE REVIEW

File: oak-run/src/main/java/org/apache/jackrabbit/oak/index/merge/IndexDiff.java [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+        // the superseded indexes of the old repository
+        List<String> supersededKeys = new ArrayList<>(IndexMerge.getSupersededIndexDefs(indexeDefinitions));
+        Collections.sort(supersededKeys);
+
+        // keep only new indexes that are not superseded
+        Map<String, JsonObject> indexMap = indexeDefinitions.getChildren();
+        for (String superseded : supersededKeys) {
+            if (indexMap.containsKey(superseded)) {
+                indexMap.remove(superseded);
+            }
```
Comment (fabriziofortino) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/index/merge/IndexDiff.java]: the `containsKey` check could be removed

```suggestion
            indexMap.remove(superseded);
```

File: oak-run/src/main/java/org/apache/jackrabbit/oak/index/merge/IndexDiff.java [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+        }
+        Set<String> indexKeys = indexeDefinitions.getChildren().keySet();
+        try {
+            IndexDefMergerUtils.merge(newIndexes, indexeDefinitions);
+            Set<String> newIndexKeys =  new HashSet<>(newIndexes.getChildren().keySet());
+            newIndexKeys.removeAll(indexKeys);
+            if (newIndexKeys.isEmpty()) {
+                // No indexes to merge
+            }
```
Comment (fabriziofortino) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/index/merge/IndexDiff.java]: this `if` is empty. Is this needed?

File: oak-run/src/main/java/org/apache/jackrabbit/oak/index/merge/IndexDiff.java [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+                    child.getProperties().put(p, v);
+                } else if (v.startsWith("\":blobId:")) {
+                    String base64 = v.substring(9, v.length() - 1);
+                    try {
+                        String clear = new String(java.util.Base64.getDecoder().decode(base64), "UTF-8");
+                        v = JsopBuilder.encode(clear);
+                        child.getProperties().put(p, v);
+                    } catch (UnsupportedEncodingException e) {
+                        throw new IllegalArgumentException(e);
+                    }
```
Comment (fabriziofortino) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/index/merge/IndexDiff.java]: `StandardCharsets.UTF_8` can be used. No need to try-catch `UnsupportedEncodingException `

```suggestion
                    String clear = new String(java.util.Base64.getDecoder().decode(base64), StandardCharsets.UTF_8);
                    v = JsopBuilder.encode(clear);
                    child.getProperties().put(p, v);
```

File: oak-run/src/main/java/org/apache/jackrabbit/oak/index/merge/IndexDiff.java [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+
+    public static void extractAll(String extractFile, String extractTargetDirectory) {
+        new File(extractTargetDirectory).mkdirs();
+        JsonObject indexDefs = parseIndexDefinitions(extractFile);
+        removeUninterestingIndexProperties(indexDefs);
+        sortPropertiesByName(indexDefs);
+        for (String child : indexDefs.getChildren().keySet()) {
+            JsonObject index = indexDefs.getChildren().get(child);
+            simplifyForDisplay(index);
+            String fileName = child.replaceAll("/oak:index/", "");
```
Comment (fabriziofortino) [inline oak-run/src/main/java/org/apache/jackrabbit/oak/index/merge/IndexDiff.java]: `"/oak:index/"` is widely used. I would create a constant.

**Review by fabriziofortino — APPROVED** [review]
LGTM. Just a few minor comments.

File: oak-run/src/main/java/org/apache/jackrabbit/oak/index/merge/IndexDiff.java (hunk 1 above)
Comment (thomasmueller) [reply oak-run/src/main/java/org/apache/jackrabbit/oak/index/merge/IndexDiff.java]: +1

**Review by fabriziofortino — APPROVED** [review]

**Comment by mreutegg** [comment]
Changes have been merged to SVN.

**Pull request closed without merging**